| MGE   |   No. Genes | Reference Genome      | Gene and Gene Position  | Associated With       | Possible Selective Function                                                                                                                                                                                                                                                                                                                                    |
|:------|:------------|:----------------------|:------------------------|:----------------------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| H1    |           3 | FMC_2547              | group_3952 - scaffold_1: 2121661..2122833 <br> group_3953 - scaffold_1: 2122925..2123291 <br> group_1459 - scaffold_1: 2124770..2125045| Human                 | Element is three hypothetical proteins, including one with a helix-hairpin-helix DNA binding domain, which might play a role in gene regulation involved in adapting to a human host. In the reference genome, genes are in separate operons with other hypothetical proteins. Adjacent genes include a putative MGE containing the virulence factor tcpC [1]. |
| H2    |           2 | FMC_2547              | | Human                 | Needs further study. Element contains two putative phage tail proteins. In the reference genome, these are in an operon with many other phage-related proteins, potentially a prophage. The specific association with human infections is unclear.                                                                                                             |
| H3    |          20 | FMC_2624              | | Human                 | Element contains all but two genes (clbK and clbQ) from the colibactin biosynthesis cluster, a known virulence factor for UPEC [2,3], along with three additional nearby genes.                                                                                                                                                                                |
| H4    |           2 | FMC_2547              | | Human                 | Needs further study. Contains two conserved hypothetical proteins. Other nearby genes include a putative MobA family protein [4], replication initiation protein, and a third hypothetical protein.                                                                                                                                                            |
| H5    |           2 | FMC_2547              | | Human                 | Contains a putative MbeB mobilization protein [4] and a hypothetical protein. Nearby are MbeC and rop/rom genes, which regulate plasmid copy number [5].                                                                                                                                                                                                       |
| H6    |          11 | FMC_1856              | | Human                 | Cluster includes membrane and periplasmic proteins, including predicted iron and heme transporters [6,7], and a scsC-like copper resistance protein [8,9].                                                                                                                                                                                                     |
| M1    |          11 | FMC_2547              | | Meat (Pork & Chicken) | Cluster of phage-related genes. In the reference genome, interrupted by additional phage-related genes, potentially forming a complete prophage. Element splits the lexA repressor [10] from recN [11], possibly affecting SOS response and DNA repair; followed by PasTI toxin-antitoxin pair [12].                                                           |
| M2    |          10 | 57_M1                 | | Meat (Chicken)        | Cluster of phage-related genes. In the reference genome, interrupted by additional phage-related genes, potentially forming a prophage. Element splits xerC and xerD, which may affect gene expression and DNA damage repair [15].                                                                                                                             |
| M3    |          42 | FMC_2624              | | Meat (Poultry)        | Conjugation machinery [16]—enables horizontal gene transfer. In the reference genome, adjacent genes include more conjugation-related genes.                                                                                                                                                                                                                   |
| M4    |          27 | 57_M1                 | | Meat (Poultry)        | Located on the ColV plasmid [17,18]. Includes colicin E1 and immunity protein [19], VapBC [20], ArtAB [21] toxin-antitoxins [12], mdtH efflux pump [22,23], and iron/heme transport proteins [17]. Genes are dispersed on the plasmid.                                                                                                                         |
| M5    |          15 | 57_M1                 | | Meat (Poultry)        | Located on the ColV plasmid [17,18]. Includes a putative efflux pump (MacAB-CusC-like) [24,25], a Mig-14-family protein [26], and the colicin V gene [27]. Genes are dispersed on the plasmid.                                                                                                                                                                 |
| M6    |           3 | 57_M1                 | | Meat (Poultry)        | Located on the ColBM plasmid [28]. Element is a continuous cluster of immunity proteins for colicin M and B and colicin M [27]. Within a larger toxin-antitoxin cluster that includes colicin B [12].                                                                                                                                                          |
| M7    |           4 | 57_M1                 | | Meat (Chicken)        | Located on the ColBM plasmid [28]. Continuous cluster of streptomycin and aminoglycoside resistance genes and groEL/groES chaperone [29].                                                                                                                                                                                                                      |
| M8    |          19 | 100_CN_05_B6_M2_C4_P2 | | Meat (Turkey)         | Copper- and silver-resistance gene cluster [8]. In the reference genome, additional resistance genes are interspersed.                                                                                                                                                                                                                                         |
| M9    |           2 | 100_CN_05_B6_M2_C4_P2 | | Meat (Turkey)         | Pair of transposases. In the reference genome, separated by another transposase and recombinase, near P fimbriae papC [6] and two neomycin resistance genes—potential impact on gene regulation.                                                                                                                                                               |
| M10   |           2 | FMC_2624              | | Meat (Turkey)         | Contains a LysR-type transcriptional regulator (controls infection-related genes [13]) and sodium-glutamate symporter (stress response [14]).                                                                                                                                                                                                                  |
| M11   |           6 | 100_CN_05_B6_M2_C4_P2 | | Meat (Turkey)         | Transposable element, adjacent to EL41.                                                                                                                                                                                                                                                                                                                        |

## References

1. Ou Q, Fang J-Q, Zhang Z-S, Chi Z, Fang J, Xu D-Y, et al. TcpC inhibits neutrophil extracellular trap formation by enhancing ubiquitination mediated degradation of peptidylarginine deiminase 4. *Nat Commun.* 2021;12:3481.

2. Morgan RN, Saleh SE, Farrag HA, Aboulwafa MM. Prevalence and pathologic effects of colibactin and cytotoxic necrotizing factor-1 (Cnf 1) in *Escherichia coli*: experimental and bioinformatics analyses. *Gut Pathog.* 2019;11:22.

3. Faïs T, Delmas J, Barnich N, Bonnet R, Dalmasso G. Colibactin: More Than a New Bacterial Toxin. *Toxins.* 2018;10. [http://dx.doi.org/10.3390/toxins10040151](http://dx.doi.org/10.3390/toxins10040151)

4. Francia MV, Varsaki A, Garcillán-Barcia MP, Latorre A, Drainas C, de la Cruz F. A classification scheme for mobilization regions of bacterial plasmids. *FEMS Microbiol Rev.* 2004;28:79–100.

5. Tomizawa J. Control of ColE1 plasmid replication. Interaction of Rom protein with an unstable complex formed by RNA I and RNA II. *J Mol Biol.* 1990;212:695–708.

6. Sarowska J, Futoma-Koloch B, Jama-Kmiecik A, et al. Virulence factors, prevalence and potential transmission of extraintestinal pathogenic *Escherichia coli* isolated from different sources. *Gut Pathog.* 2019. [http://dx.doi.org/10.1186/s13099-019-0290-0](http://dx.doi.org/10.1186/s13099-019-0290-0)

7. Subashchandrabose S, Mobley HLT. Virulence and Fitness Determinants of Uropathogenic *Escherichia coli*. *Microbiol Spectr.* 2015;3. [http://dx.doi.org/10.1128/microbiolspec.UTI-0015-2012](http://dx.doi.org/10.1128/microbiolspec.UTI-0015-2012)

8. Cheng G, Ning J, Ahmed S, et al. Selection and dissemination of antimicrobial resistance in Agri-food production. *Antimicrob Resist Infect Control.* 2019. [http://dx.doi.org/10.1186/s13756-019-0623-2](http://dx.doi.org/10.1186/s13756-019-0623-2)

9. Shepherd M, Heras B, Achard MES, et al. Structural and functional characterization of ScsC, a periplasmic thioredoxin-like protein from *Salmonella enterica* serovar Typhimurium. *Antioxid Redox Signal.* 2013;19:1494–506.

10. Butala M, Zgur-Bertok D, Busby SJW. The bacterial LexA transcriptional repressor. *Cell Mol Life Sci.* 2009;66:82–93.

11. Keyamura K, Hishida T. Topological DNA-binding of structural maintenance of chromosomes-like RecN promotes DNA double-strand break repair in *Escherichia coli*. *Commun Biol.* 2019;2:413.

12. Norton JP, Mulvey MA. Toxin-antitoxin systems are important for niche-specific colonization and stress resistance of uropathogenic *Escherichia coli*. *PLoS Pathog.* 2012;8:e1002954.

13. Maddocks SE, Oyston PCF. Structure and function of the LysR-type transcriptional regulator (LTTR) family proteins. *Microbiology.* 2008;154:3609–23.

14. Feehily C, Karatzas KAG. Role of glutamate metabolism in bacterial responses towards acid and other stresses. *J Appl Microbiol.* 2013;114:11–24.

15. Rudenko G, Hohenester E, Muller YA. LG/LNS domains: multiple functions – one business end? *Trends Biochem Sci.* 2001;26:363–8. [http://dx.doi.org/10.1016/s0968-0004(01)01832-1](http://dx.doi.org/10.1016/s0968-0004(01)01832-1)

16. Cabezón E, Ripoll-Rozada J, Peña A, de la Cruz F, Arechaga I. Towards an integrated model of bacterial conjugation. *FEMS Microbiol Rev.* 2015;39:81–95.

17. Johnson TJ, Siek KE, Johnson SJ, Nolan LK. DNA sequence of a ColV plasmid and prevalence of selected plasmid-encoded virulence genes among avian *Escherichia coli* strains. *J Bacteriol.* 2006;188:745–58.

18. Johnson TJ, Logue CM, Wannemuehler Y, et al. Examination of the source and extended virulence genotypes of *Escherichia coli* contaminating retail poultry meat. *Foodborne Pathog Dis.* 2009;6:657–67.

19. Duché D, Issouf M, Lloubès R. Immunity protein protects colicin E2 from OmpT protease. *J Biochem.* 2009;145:95–101.

20. Winther KS, Gerdes K. Ectopic production of VapCs from Enterobacteria inhibits translation and trans-activates YoeB mRNA interferase. *Mol Microbiol.* 2009;72:918–30.

21. Littler DR, Ang SY, Moriel DG, et al. Structure–function analyses of a pertussis-like toxin from pathogenic *Escherichia coli*. *J Biol Chem.* 2017;292:15143–58.

22. Zhang Y, Dong R, Zhang M, Gao H. Native efflux pumps of *Escherichia coli* responsible for short and medium chain alcohol. *Biochem Eng J.* 2018;133:149–56.

23. Yu L, Li W, Li Q, et al. Role of LsrR in the regulation of antibiotic sensitivity in avian pathogenic *Escherichia coli*. *Poult Sci.* 2020;99:3675–87.

24. Souabni H, Batista Dos Santos W, Cece Q, et al. Quantitative real-time analysis of the efflux by the MacAB-TolC tripartite efflux pump. *Commun Biol.* 2021;4:493.

25. Chacón KN, Mealman TD, McEvoy MM, Blackburn NJ. Tracking metal ions through a Cu/Ag efflux pump. *Proc Natl Acad Sci U S A.* 2014;111:15373–8.

26. McPhee JB, Small CL, Reid-Yu SA, et al. Host defense peptide resistance contributes to colonization and intestinal pathology by *E. coli*. *Infect Immun.* 2014;82:3383–93.

27. Cascales E, Buchanan SK, Duché D, et al. Colicin Biology. *Microbiol Mol Biol Rev.* 2007;71:158–229. [http://dx.doi.org/10.1128/mmbr.00036-06](http://dx.doi.org/10.1128/mmbr.00036-06)

28. Johnson TJ, Johnson SJ, Nolan LK. Complete DNA Sequence of a ColBM Plasmid. *J Bacteriol.* 2006;188:5975–83. [http://dx.doi.org/10.1128/jb.00204-06](http://dx.doi.org/10.1128/jb.00204-06)

29. Fourie KR, Wilson HL. Understanding GroEL and DnaK Stress Response Proteins as Antigens. *Vaccines.* 2020;8:773. [http://dx.doi.org/10.3390/vaccines8040773](http://dx.doi.org/10.3390/vaccines8040773)

